src/message/messagesActions.js

```diff
@@ -47,5 +51,33 @@ export const messageLinkPress = (href: string) => (dispatch: Dispatch, getState:
     dispatch(doNarrow(narrow, anchor));
     return;
   }
-  openLink(getFullUrl(href, auth.realm));
+  const url = getFullUrl(href, auth.realm);
+
+  // We don't want to generate temporary URLs to external websites.
+  if (!url.includes(auth.realm)) {
+    openLink(url);
+    return;
+  }
+
+  let tempUrl: string;
+  try {
+    tempUrl = getFullUrl((await getTemporaryFileUrl(auth, href)).url, auth.realm);
+  } catch (err) {
+    // If we failed to generate the temporary URL, we assume it's because the
+    // server doesn't suport it yet, and fallback to use the regular file
+    // URL.
+    openLink(url);
+    return;
+  }
+
+  // We want to support opening .pdf files directly in the browser for
+  // instant viewing. But for other files, we want them to be directly
+  // downloaded. This prevents opening an ugly blank browser window that
+  // eventually automatically does the same thing - i.e. downloads the
```

Comment:
It's nice to avoid the detour into the browser, but one thing the browser does for us, I think, is make it more likely that files of varying types (not just PDFs) eventually get displayed in a nice way. Might there be other types of files, besides PDFs, that we'd want to display in the browser?

Also, if the browser decides it can't display it, it triggers what seems like the platform's preferred way to download a file. On iOS, at least, it takes me through an elaborate flow where I decide whether I want to save it to iCloud, or onto my device, and where, etc. There may be a simple way to hook into this native functionality on iOS, but currently, `downloadFile` doesn't do that — it tries to save all non-PDF files to the camera roll, and (at least on the simulator) does nothing with a `.docx` file.

@gnprice, do you think we should aim for a middle ground, and do some direct downloads (perhaps we can find out how to trigger iOS's native file-download flow?), or would it be best to just let the browser handle everything?

Reply:
The browser is a good fallback because it's sure to know how to do something sensible (to the extent there is something sensible to be done.)

Looking at the implementation of `downloadImage`, it has basically totally separate codepaths for iOS and Android. On their face the iOS codepath specifically saves to the camera roll, but the Android codepath looks like a generic file download. So it's possible that for Android we could use the logic we already have and do a direct download with that, without the browser. For iOS we'd need to figure out some other API to use.

I think that doesn't need to block fixing this issue, though, so leaving that out for now (as in the new revision of this PR) sounds like a good idea.

Reply:
(ah, and @agrawal-d I see you'd replied in the main thread! That is actually probably the better place to continue this discussion :slightly_smiling_face: -- when a conversation goes more than a round or two, it's easy for it to get lost if in the inline comments)